pkg/internal/token/clientcertcredential.go

```diff
@@ -0,0 +1,206 @@
+	}
+
+	if opts.httpClient != nil {
+		azOpts.ClientOptions.Transport = opts.httpClient
+	}
+
+	cred, err := azidentity.NewClientCertificateCredential(
+		opts.TenantID, opts.ClientID,
+		[]*x509.Certificate{cert}, rsaPrivateKey,
+		azOpts)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create client certificate credential: %s", err)
+	}
+	return &ClientCertificateCredential{cred: cred}, nil
+}
+
+func (c *ClientCertificateCredential) Name() string {
+	return "ClientCertificateCredential"
+}
+
+func (c *ClientCertificateCredential) Authenticate(ctx context.Context, opts *policy.TokenRequestOptions) (azidentity.AuthenticationRecord, error) {
+	panic("not implemented")
+}
+
+func (c *ClientCertificateCredential) GetToken(ctx context.Context, opts policy.TokenRequestOptions) (azcore.AccessToken, error) {
+	return c.cred.GetToken(ctx, opts)
+}
+
+func (c *ClientCertificateCredential) NeedAuthenticate() bool {
+	return false
+}
+
+func isPublicKeyEqual(key1, key2 *rsa.PublicKey) bool {
+	if key1.N == nil || key2.N == nil {
+		return false
+	}
+	return key1.E == key2.E && key1.N.Cmp(key2.N) == 0
+}
+
+func splitPEMBlock(pemBlock []byte) (certPEM []byte, keyPEM []byte) {
+	for {
+		var derBlock *pem.Block
+		derBlock, pemBlock = pem.Decode(pemBlock)
+		if derBlock == nil {
+			break
+		}
+		if derBlock.Type == "CERTIFICATE" {
+			certPEM = append(certPEM, pem.EncodeToMemory(derBlock)...)
+		} else if derBlock.Type == "PRIVATE KEY" {
+			keyPEM = append(keyPEM, pem.EncodeToMemory(derBlock)...)
+		}
+	}
+
+	return certPEM, keyPEM
+}
+
+func parseRsaPrivateKey(privateKeyPEM []byte) (*rsa.PrivateKey, error) {
+	block, _ := pem.Decode(privateKeyPEM)
+	if block == nil {
+		return nil, fmt.Errorf("failed to decode a pem block from private key")
+	}
+
+	privatePkcs1Key, errPkcs1 := x509.ParsePKCS1PrivateKey(block.Bytes)
+	if errPkcs1 == nil {
+		return privatePkcs1Key, nil
+	}
+
+	privatePkcs8Key, errPkcs8 := x509.ParsePKCS8PrivateKey(block.Bytes)
+	if errPkcs8 == nil {
+		privatePkcs8RsaKey, ok := privatePkcs8Key.(*rsa.PrivateKey)
+		if !ok {
+			return nil, fmt.Errorf("pkcs8 contained non-RSA key. Expected RSA key")
+		}
+		return privatePkcs8RsaKey, nil
+	}
+
+	return nil, fmt.Errorf("failed to parse private key as Pkcs#1 or Pkcs#8. (%s). (%s)", errPkcs1, errPkcs8)
+}
+
+func parseKeyPairFromPEMBlock(pemBlock []byte) (*x509.Certificate, *rsa.PrivateKey, error) {
+	certPEM, keyPEM := splitPEMBlock(pemBlock)
+
+	privateKey, err := parseRsaPrivateKey(keyPEM)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	found := false
+	var cert *x509.Certificate
+	for {
+		var certBlock *pem.Block
+		var err error
+		certBlock, certPEM = pem.Decode(certPEM)
+		if certBlock == nil {
+			break
+		}
+
+		cert, err = x509.ParseCertificate(certBlock.Bytes)
+		if err != nil {
+			return nil, nil, fmt.Errorf("unable to parse certificate. %w", err)
+		}
+
+		certPublicKey, ok := cert.PublicKey.(*rsa.PublicKey)
+		if ok {
```

Comment:
nit: how about we combine the nested if checks together?

```
if ok && isPublicKeyEqual(certPublicKey, &privateKey.PublicKey) {
  found = true
  break
}
```

Reply:
fixed in next commit